src/mds/MDSMap.cc

```diff
@@ -777,94 +735,45 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   }
 
   // kclient ignores everything from here
-  __u16 ev = 1;
   if (struct_v >= 2)
     decode(ev, p);
+  if (decode_ev)
+    ceph_assert(ev == 5);
   if (ev >= 3)
     decode(compat, p);
   else
     compat = get_compat_set_base();
-  if (ev < 5) {
-    __u32 n;
-    decode(n, p);
-    metadata_pool = n;
-  } else {
-    decode(metadata_pool, p);
-  }
+
   decode(created, p);
   decode(modified, p);
-  decode(tableserver, p);
   decode(in, p);
   decode(inc, p);
   decode(up, p);
   decode(failed, p);
   decode(stopped, p);
-  if (ev >= 4)
+  if (ev >= 5)
     decode(last_failure_osd_epoch, p);
-  if (ev >= 6) {
-    if (ev < 10) {
-      // previously this was a bool about snaps, not a flag map
-      bool flag;
-      decode(flag, p);
-      ever_allowed_features = flag ? CEPH_MDSMAP_ALLOW_SNAPS : 0;
-      decode(flag, p);
-      explicitly_allowed_features = flag ? CEPH_MDSMAP_ALLOW_SNAPS : 0;
-    } else {
+  DECODE_FINISH(p);
+
+  {
+    // ev part
+    if (decode_ev)
+    {
+      DECODE_START(1, p);
+      decode(metadata_pool, p);
+      decode(tableserver, p);
       decode(ever_allowed_features, p);
       decode(explicitly_allowed_features, p);
+      decode(inline_data_enabled, p);
+      decode(enabled, p);
+      decode(damaged, p);
+      decode(balancer, p);
+      decode(standby_count_wanted, p);
+      decode(old_max_mds, p);
+      decode(min_compat_client, p);
+      DECODE_FINISH(p);
     }
-  } else {
-    ever_allowed_features = 0;
-    explicitly_allowed_features = 0;
   }
-  if (ev >= 7)
-    decode(inline_data_enabled, p);
-
-  if (ev >= 8) {
-    ceph_assert(struct_v >= 5);
-    decode(enabled, p);
-    decode(fs_name, p);
-  } else {
-    if (epoch > 1) {
-      // If an MDS has ever been started, epoch will be greater than 1,
-      // assume filesystem is enabled.
-      enabled = true;
-    } else {
-      // Upgrading from a cluster that never used an MDS, switch off
-      // filesystem until it's explicitly enabled.
-      enabled = false;
-    }
-  }
-
-  if (ev >= 9) {
-    decode(damaged, p);
-  }
-
-  if (ev >= 11) {
-    decode(balancer, p);
-  }
-
-  if (ev >= 12) {
-    decode(standby_count_wanted, p);
-  }
-
-  if (ev >= 13) {
-    decode(old_max_mds, p);
-  }
-
-  if (ev == 14) {
-    int8_t r;
-    decode(r, p);
-    if (r < 0) {
-      min_compat_client = ceph_release_t::unknown;
-    } else {
-      min_compat_client = ceph_release_t{static_cast<uint8_t>(r)};
-    }
-  } else if (ev > 14) {
-    decode(min_compat_client, p);
-  }
```

Comment:
you need to keep decode code for ev in range 6 ~ 14

Reply:
@ukernel I reverted the range 6-14. Please check. I'm doubtful if the decode ev part is correct. It looks like there are 2 ev parts now.

Reply:
> > ```
> > * with this change. it's better to let monitor only encode full mdsmap for mds, skip the new extented part for clients
> > ```
> 
> @ukernel
> I think you meant the entire ev part (everything below `encode(ev, bl);`)? Putting the new ev part alone in a condition fails the test during make check.

client does not need fields from 'compat to min_compat_client'. it's better to add a parameter to MDSMap message 's constructor. if message is for client, skip encoding them

Reply:
@ukernel I have pushed a sample of the client checking for now. We may need to expand it.

I'm not so happy with the fix as the MDSMap object can't be modified to update `is_client`. Could you please take a look?